src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs

```diff
@@ -384,36 +389,51 @@ private int ParseMultiSpanHeader(TRequestHandler handler, ref SequenceReader<byt
             header = currentSlice.Slice(reader.Position, lineEnd);
         }
 
-        var headerSpan = header.ToSpan();
+        byte[]? array = null;
+        int length = (int)header.Length;
+        Span<byte> span = length <= 256 ? stackalloc byte[length] : array = ArrayPool<byte>.Shared.Rent(length);
 
-        // 'a:b\n' or 'a:b\r\n'
-        var minHeaderSpan = _disableHttp1LineFeedTerminators ? 5 : 4;
-        if (headerSpan.Length < minHeaderSpan)
+        try
         {
-            RejectRequestHeader(headerSpan);
-        }
+            header.CopyTo(span);
+            span = span.Slice(0, length);
 
-        var terminatorSize = -1;
+            // 'a:b\n' or 'a:b\r\n'
+            var minHeaderSpan = _disableHttp1LineFeedTerminators ? 5 : 4;
+            if (span.Length < minHeaderSpan)
+            {
+                RejectRequestHeader(span);
+            }
 
-        if (headerSpan[^1] == ByteLF)
-        {
-            if (headerSpan[^2] == ByteCR)
+            var terminatorSize = -1;
+
+            if (span[^1] == ByteLF)
             {
-                terminatorSize = 2;
+                if (span[^2] == ByteCR)
+                {
+                    terminatorSize = 2;
+                }
+                else if (!_disableHttp1LineFeedTerminators)
+                {
+                    terminatorSize = 1;
+                }
             }
-            else if (!_disableHttp1LineFeedTerminators)
+
+            // Last chance to bail if the terminator size is not valid or the header doesn't parse.
+            if (terminatorSize == -1 || !TryTakeSingleHeader(handler, span.Slice(0, span.Length - terminatorSize)))
             {
-                terminatorSize = 1;
+                RejectRequestHeader(span);
             }
-        }
 
-        // Last chance to bail if the terminator size is not valid or the header doesn't parse.
-        if (terminatorSize == -1 || !TryTakeSingleHeader(handler, headerSpan.Slice(0, headerSpan.Length - terminatorSize)))
+            return span.Length;
+        }
+        finally
```

Comment:
The `try-finally` shouldn't be used, just return the rented array, and in case of exception let the GC do it's job.

More details and reasoning in https://github.com/dotnet/runtime/issues/71249#issuecomment-1165365583 ff.

Reply:
David said the same thing to me offline 😄

Reply:
@gfoidl  @BrennanConroy   that can cause memory leak, and then cause the denial-of-service , the GC won't release the pool
![image](https://github.com/dotnet/aspnetcore/assets/13639146/f96c9fea-d046-4352-b358-0e8f49df50ac)
and after another 20 `TakeThePool`
![image](https://github.com/dotnet/aspnetcore/assets/13639146/4ae1e1e3-b8eb-4f10-a69b-0f8b39af361b)

it won't run into this code  in normal http protocol , but what if it's a http like TCP stream ?

Reply:
@John0King can you please provide a repro so we can have a look into it?
The screenshots are nice, but hides what your code (`TakeThePool`) is doing (w/ the ArrayPool).

Returning from the ArrayPool shouldn't leak, wheter it's in a try-finally or not.
If there's just renting (and the rented memory is rooted) then of course the GC has to keep that memory alive.

Reply:
@gfoidl  sorry, I miss that code in the screenshots 

```c#
int TakeThePool()
{
      var arr = System.Buffers.ArrayPool<byte>.Shared.Rent(40_000_000);
      return arr.Length; // make sure the firstline is been executed;
}
```

 I'm arguing that  throw without `Return` to the `ArrayPool` will never release the pool, so every times it happened that will case the process memory bigger and bigger untill OOM exception .   
 
 I'm doing the same test today and the memory is not grow , I can't reproduce the same result as yesterday.  it seems the ArrayPool is doing hard work to allow GC to collect the  rented array

Reply:
Note that renting an array of length 40_000_000 will be actually an LOH allocation (large object heap, default threshould is 85_000 bytes object size).

I'm tempted to say that renting such big objects from the ArrayPool is mis-use, as the pool isn't designed for such big objects.